### PR TITLE
Fix read "max" value for cgroup v2

### PR DESCRIPTION
**Pull request by giuseppe**
and also simplify some other code to behave like cgroup v1

Signed-off-by: Giuseppe Scrivano <gscrivan@redhat.com>

### CODE REVIEW

**Comment by giuseppe** [comment]
@odinuge PTAL

File: container/common/helpers.go [hunk 1]
```diff
@@ -165,11 +146,7 @@ func GetSpec(cgroupPaths map[string]string, machineInfoFactory info.MachineInfoF
 				spec.Memory.Reservation = readUInt64(memoryRoot, "memory.soft_limit_in_bytes")
 			}
 		} else {
-			memoryRoot, err := findFileInAncestorDir(memoryRoot, "memory.max", "/sys/fs/cgroup")
-			if err != nil {
-				return spec, err
-			}
-			if memoryRoot != "" {
+			if utils.FileExists(path.Join(memoryRoot, "memory.max")) {
```
Comment (odinuge) [inline container/common/helpers.go:149]: :+1:

File: container/common/helpers.go [hunk 2]
```diff
@@ -226,7 +203,10 @@ func readString(dirpath string, file string) string {
 
 func readUInt64(dirpath string, file string) uint64 {
 	out := readString(dirpath, file)
-	if out == "" || out == "max" {
+	if out == "max" {
+		return math.MaxUint64
```
Comment (odinuge) [inline container/common/helpers.go:207]: nit: The max value for cgroup v1 is 9223372036854771712 (max signed integer rounded to nearest page). So to be correct it should be something like this I guess:

```suggestion
		return math.MaxInt64^0xFFF
```

**Review by odinuge — COMMENTED** [review]
Just a small nit, otherwise lgtm

File: container/common/helpers.go (hunk 2 above)
Comment (giuseppe) [reply container/common/helpers.go:207]: I am fine with changing it, but would it be a problem with cgroup v1?  `"max"` is used only on cgroup v2

Comment (odinuge) [reply container/common/helpers.go:207]: I don't remember where this value is used, but I just thought it was nice with the same value for both cgroup v2. If you try to write `MaxInt64^0xFFF  + 0x1` you get `max` on cgroup v2, and `9223372036854771712` on v1. 9223372036854771712 is also default on v1.

Comment (odinuge) [reply container/common/helpers.go:207]: Edit: This is used for more than the memory limits, and my above comments only apply to those. Hmm

**Comment by giuseppe** [comment]
a similar fix for cpu specs: https://github.com/google/cadvisor/pull/2839

File: container/common/helpers.go [hunk 3]
```diff
@@ -226,7 +203,10 @@ func readString(dirpath string, file string) string {
 
 func readUInt64(dirpath string, file string) uint64 {
 	out := readString(dirpath, file)
-	if out == "" || out == "max" {
+	if out == "max" {
```
Comment (iwankgb) [inline container/common/helpers.go:206]: Why have you decided to change behaviour for `out == ""`?

Comment (giuseppe) [reply container/common/helpers.go:206]: for `out == ""` I've left the same behavior there was before 2ccad4b42fe52b312f6d75a312c61a54fd123dc5

**Comment by odinuge** [comment]
This change is probably related to:
https://github.com/google/cadvisor/blob/fa0733216494720e3fc301bdf046553f407a7131/manager/manager.go#L473-L479

Does that code do anything at all if this change merged?


Also, the description of this value is probably wrong, since it says -1 is default, but it is an unsigned value...

https://github.com/google/cadvisor/blob/fa0733216494720e3fc301bdf046553f407a7131/info/v2/container.go#L46-L48

**Comment by giuseppe** [comment]
> [](/google/cadvisor/commit/fa0733216494720e3fc301bdf046553f407a7131)

yes it will be ignored.

And I think this is the expectation, at least from Kubernetes, since there is a e2e tests failing right now as the memory limit is being set while it is expected to be unlimited (`isMemoryUnlimited` in pkg/kubelet/stats/helper.go)

File: container/common/helpers.go (hunk 3 above)
Comment (iwankgb) [reply container/common/helpers.go:206]: Of course, my bad!

**Review by iwankgb — APPROVED** [review]

**Pull request merged**